### PR TITLE
Add Weekly Digest Command

### LINKED ISSUE #509: Add Weekly Digest Command
Problem

Termstory currently does not provide a single command for users to review their development activity over the course of a week.

Users may have difficulty understanding:

How much time they spent working during the week.
Which days were the most active.
How many sessions, commands, and commits they completed.
Which projects received the most time.
What times of day they are most productive.
How their focus and activity have changed over time.
Whether they are maintaining consistent activity streaks.

Reviewing individual activity records makes it difficult to get a complete weekly overview.

Proposed Solution

Add a new termstory weekly CLI command that generates a comprehensive weekly digest from recorded activity.

The report should include:

Weekly summary statistics.
Daily session, command, and commit breakdown.
Visual activity bars.
Project time distribution.
Time-of-day activity heatmap.
Command category frequency.
Focus score.
Activity streak information.
Top achievements.
Previous Week Comparison

Add a --last option to display the previous week's digest and calculate week-over-week changes for:

Time spent.
Commits.
Sessions.
Focus score.

The comparison should include percentage changes where applicable.

Testing

The analytics logic should remain separated from database and CLI I/O so it can be tested independently.

Add comprehensive tests covering weekly calculations, aggregation, streaks, comparisons, formatting, and edge cases.

Alternatives Considered

Users could manually inspect individual activity records or combine the data in external spreadsheets, but this would be time-consuming and would not provide an immediate overview.

A basic summary containing only total time and commits would be simpler, but it would miss useful information such as daily patterns, project distribution, focus, and week-over-week changes.

A dedicated weekly digest provides a more complete and convenient way to understand development activity.

Willingness to Contribute

I would like to submit a Pull Request to implement this feature.

### PR DESCRIPTION
**Pull request by karan-chaos**
closes #509 
Summary

This PR adds a new termstory weekly CLI command that generates a comprehensive weekly productivity and activity report.

Features
Weekly Analytics

The command provides:

Weekly summary statistics.
Daily session, command, and commit breakdown.
Visual activity bars.
Project-wise time distribution.
Time-of-day activity heatmap.
Command category frequency.
Focus score.
Activity streak tracking.
Top achievements.
Previous Week Comparison

Added the --last flag to view the previous week's report.

When comparing weeks, the command automatically calculates percentage changes for:

Time spent.
Commits.
Sessions.
Focus score.
Architecture
Added a pure-computation analytics engine in weekly_digest.py.
Kept analytics logic independent of I/O for easier testing and reuse.
Added a dedicated formatter for rendering the weekly report.
Wired the new command into the existing CLI and database flow.
Testing

Added 30+ unit tests covering:

Analytics helpers.
Weekly calculations.
Activity aggregation.
Streak tracking.
Week-over-week comparisons.
Formatting and integration behaviour.
Files Changed
termstory/weekly_digest.py
termstory/formatter.py
termstory/cli.py
tests/test_weekly_digest.py
Impact

Users can now generate a detailed weekly overview of their development activity and productivity directly from the CLI, while the --last option makes it easy to compare current performance with the previous week.

Branch: feature/weekly-digest-command

Commit: 6eb32d9 — feat(cli): add weekly digest command with comprehensive analytics

### CODE REVIEW

File: termstory/cli.py [hunk 1]
```diff
@@ -555,6 +555,48 @@ def show_tags(
...
+    safe_init_db(db)
+
+    run_ingestion(db)
+
+    start_ts, end_ts = get_week_range(last=last)
+    sessions = db.get_range_sessions(start_ts, end_ts)
+
+    # Previous-week sessions for comparison
+    prev_start, prev_end = get_week_range(last=True)
+    previous_sessions = db.get_range_sessions(prev_start, prev_end)
```
Comment (bitflicker64) [inline termstory/cli.py:574]: `get_week_range(last=True)` always loads the previous week. When the user already passed `--last`, both `sessions` and `previous_sessions` are the same window, so the comparison is week-vs-self. Need a range that is one week before the selected week.

**Review by bitflicker64 — COMMENTED** [review]
Summary: adds `termstory weekly` plus weekly_digest.py and formatter output. Linked issue #509 is in the right area.

Needs a manual look. First merged PR from this author, only Greptile ran (no Test/Lint jobs), mergeable_state=unstable, and `--last` compares a week to itself. Not merging.

CI: Greptile success. No Test (Python 3.9-3.12) or Lint runs on this head.

**Comment by bitflicker64** [comment]
Needs a manual look. First contribution from this author, required Test/Lint jobs did not run on this head, and `--last` compares a week to itself. Not merging from this pass.

File: termstory/cli.py [hunk 2]
```diff
@@ -555,6 +555,48 @@ def show_tags(
...
+    db_path = get_db_path()
+    db = Database(db_path)
+    safe_init_db(db)
+
+    run_ingestion(db)
+
+    start_ts, end_ts = get_week_range(last=last)
+    sessions = db.get_range_sessions(start_ts, end_ts)
+
+    # Previous-week sessions for comparison
```
Comment (bitflicker64) [inline termstory/cli.py:572]: `get_week_range(last=True)` is always used for the comparison window. When the user already passed `--last`, current and previous become the same week and the deltas are all zero. Offset one more week when `last` is set.

File: termstory/weekly_digest.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  all Rich rendering so this module stays testable.
+* Operates on raw SQLite rows via the Database helper methods (same
+  pattern as ``insights.py``).
+* All public helpers accept plain data, not DB connections, so unit tests
+  can exercise them with hand-crafted inputs.
+"""
+
+from __future__ import annotations
+
+import os
```
Comment (bitflicker64) [inline termstory/weekly_digest.py:20]: `import os` is unused. Drop it.

**Review by bitflicker64 — COMMENTED** [review]
Summary: new `termstory weekly` command plus tests, aimed at #509. Core split of compute vs format is fine.

`--last` compares the week against itself because comparison always calls `get_week_range(last=True)`. Unused `os` import. Test matrix did not run (Greptile only). Needs a manual look: first contribution from this author and ~1000 new lines.

Does not merge as-is.

**Comment by bitflicker64** [comment]
Needs a manual look before merge. First contribution from this author, large new CLI surface, and the test matrix did not run on this fork PR.

**Review by bitflicker64 — COMMENTED** [review]
Summary: large feature adding weekly_digest and CLI weekly command.

Needs a manual look. Author has no previously merged PR in this repo. Full Test CI not present. Label do-not-merge/needs-star is present. Escalating.

**Comment by bitflicker64** [comment]
Needs a manual look before any merge decision.

**Review by bitflicker64 — CHANGES_REQUESTED** [review]
Summary: adds termstory weekly plus weekly_digest.py, formatter output, and a large unit test file. That matches #509 at a high level.

Does not land as-is.

termstory/cli.py (show_weekly): prev_start, prev_end = get_week_range(last=True) always loads last week. When --last is already set, current and previous ranges are the same week, so the comparison is a no-op. Compute the prior window from start_ts instead.

termstory/weekly_digest.py (_daily_session_totals and siblings): datetime.fromtimestamp(s.start_time) will raise on out-of-range timestamps. Other stats paths use _safe_datetime_from_ts and skip those rows.

CI: only Greptile Review ran; no Test (Python 3.9-3.12) or Lint jobs on this head. mergeable_state=unstable. Label do-not-merge/needs-star is present.

Needs a manual look before merge. Holding for that plus the --last comparison bug.

**Comment by bitflicker64** [comment]
Needs a manual look before merge. First-time contributor on this repo, ~1000 new lines, and no Test CI jobs on the head commit.

File: termstory/cli.py (hunk 1 above)
Comment (bitflicker64) [inline termstory/cli.py:574]: `get_week_range(last=True)` is always the previous week, even when the command itself was invoked with `--last`. In that case current and previous are the same window, so week-over-week deltas are zero. Compute the comparison week relative to the displayed week.

**Comment by bitflicker64** [comment]
Large new feature adding weekly_digest.py and weekly command. Author has no prior merged PRs in this repo. Label do-not-merge/needs-star is present. mergeable_state is unstable and full CI test matrix is not green. Holding for now.

File: termstory/weekly_digest.py (hunk 3 above)
Comment (bitflicker64) [inline termstory/weekly_digest.py:20]: `os` is unused. Drop the import.

**Review by bitflicker64 — CHANGES_REQUESTED** [review]
This is the weekly digest asked for in #509, but comparison is wrong when `--last` is set: both windows come from `get_week_range(last=True)`, so deltas are always zero for last week.

Also drop the unused `os` import in `weekly_digest.py`. Comparison week must be the week before the displayed week.

**Review by bitflicker64 — CHANGES_REQUESTED** [review]
Summary: this adds `termstory weekly` against #509, with a new `weekly_digest.py`, formatter, CLI wiring, and a large test file. That matches the linked issue at a high level, but the comparison window is wrong and required CI never ran (only Greptile Review is present). Also first merged-PR from this author is not on record here, so this stays out of the merge queue.

termstory/cli.py show_weekly: `get_week_range(last=last)` selects the displayed week, then `get_week_range(last=True)` is always used for the baseline. With `--last`, current and previous are the same window, so every delta is zero. When `last=True` the baseline must be the week before that window.

termstory/weekly_digest.py `_top_achievements`: command highlights copy `cmd.command` straight into the digest. CONTRIBUTING.md requires `redact_command()` / `sanitize_session_commands()` before command text leaves the process. Notebook already does this.

termstory/formatter.py `format_weekly_digest_output`: several `datetime.fromtimestamp(...)` calls have no guard. That will crash the new command on the same bad rows #464 / #506 already track.

Do not merge.

**Review by bitflicker64 — CHANGES_REQUESTED** [review]
Summary: adds termstory weekly plus weekly_digest.py, formatter output, and tests. That matches #509 at a high level.

termstory/cli.py show_weekly: get_week_range(last=last) selects the displayed week, then get_week_range(last=True) is always used for the baseline. With --last, current and previous are the same window, so every delta is zero. When last=True the baseline must be the week before that window.

termstory/weekly_digest.py _top_achievements: command highlights copy cmd.command straight into the digest. Redact before command text leaves the process.

termstory/weekly_digest.py helpers: datetime.fromtimestamp(s.start_time) has no guard. A bad timestamp will crash the new command.

CI: only Greptile Review on this head. No Test or Lint jobs. mergeable_state=unstable. Label do-not-merge/needs-star.

First merged PR from this author is not on record here. Fix the --last comparison window. Do not merge.
